templates/vm.xml.j2

```diff
@@ -1,38 +1,45 @@
 <domain type='{{ libvirt_vm_engine }}'>
-  <name>{{ libvirt_vm_name }}</name>
-  <memory>{{ libvirt_vm_memory_mb | int * 1024 }}</memory>
-  <vcpu>{{ libvirt_vm_vcpus }}</vcpu>
+  <name>{{ vm.name }}</name>
+  <memory>{{ vm.memory_mb | int * 1024 }}</memory>
+  <vcpu>{{ vm.vcpus }}</vcpu>
   <clock sync="localtime"/>
+  <on_poweroff>destroy</on_poweroff>
+  <on_reboot>restart</on_reboot>
+  <on_crash>destroy</on_crash>
   <os>
-    <type arch='{{ libvirt_vm_arch }}'{% if libvirt_vm_machine is not none %} machine='{{ libvirt_vm_machine }}'{% endif %}>hvm</type>
+    <type arch='{{ libvirt_vm_arch }}'{% if vm.machine is defined %} machine='{{ vm.machine | default('pc-1.0') }}'{% endif %}>hvm</type>
     <bootmenu enable='no'/>
+    <boot dev='hd'/>
+    <boot dev='network'/>
     <bios useserial='yes'/>
   </os>
-  <cpu{% if libvirt_vm_cpu_mode is not none %} mode='{{ libvirt_vm_cpu_mode }}'{% endif %}>
+  <cpu mode='{% if vm.cpu_mode is defined and vm.cpu_mode is not none %}{{ vm.cpu_mode }}{% elif libvirt_vm_engine == 'kvm' %}host_passthrough{% else %}host-model{% endif %}'>
     <model fallback='allow'/>
   </cpu>
   <devices>
     <emulator>{{ libvirt_vm_emulator }}</emulator>
-{% for volume in libvirt_vm_volumes %}
+{% for volume in (vm.volumes | default([])) %}
     <disk type='volume' device='{{ volume.device | default('disk') }}'>
       <driver name='qemu' type='{{ volume.format }}'/>
       <source pool='{{ volume.pool }}' volume='{{ volume.name }}'/>
       <target dev='vd{{ 'abcdefghijklmnopqrstuvwxyz'[loop.index] }}'/>
     </disk>
 {% endfor %}
-{% for interface in libvirt_vm_interfaces %}
+{% for interface in (vm.interfaces | default([])) %}
     <interface type='network'>
       <source network='{{ interface.network }}'/>
       <model type='virtio'/>
     </interface>
 {% endfor %}
      <serial type='file'>
-       <source path='{{ libvirt_vm_console_log_path }}'/>
+       <source path='{{ vm.console_log_path | default('/var/log/libvirt-consoles/' + vm.name + '.log') }}'/>
      </serial>
      <serial type='pty'/>
      <console type='file'>
-       <source path='{{ libvirt_vm_console_log_path }}'/>
+       <source path='{{ vm.console_log_path | default('/var/log/libvirt-consoles/' + vm.name + '.log') }}'/>
```

Comment:
If you wanted to avoid repeating this logic, you could define a per-task variable. You could even do something like this in main.yml:

```
- include_tasks: create_vm.yml
  vars:
    vm_console_path: "{{ vm.console_log_path | default('/var/log/libvirt-consoles/' + vm.name + '.log') }}"
    ...
```

Reply:
Can't believe I didn't think of that - much better idea.